Create chain tx

kmenon01 opened the pull request:
Finished working implementation of Create Chain Transaction

Commit:
Finished initial implementation of create chain tx, moving on to testing

kmenon01 committed:
fixed compilation issues and now moving to configure testing

Commit:
added hooks for Create Chain parsing and modified test

kmenon01 committed:
Create Chain is parsing properly, need to figure out guideline for prompts

Commit:
changed up logic for genesis data parsing and chain name parsing

kmenon01 committed:
successfully parsing create chain spec, will need to modify for correct headers

kmenon01 committed:
Merge branch 'develop' into create_chain_tx

kmenon01 committed:
Merge branch 'develop' into create_chain_tx

kmenon01 committed:
merged in develop to include add_sv_fix

kmenon01 committed:
Merge branch 'develop' into create_chain_tx

kmenon01 committed:
Parsing a portion of the create chain transaction with the correct prompts, currently trying to resolve bug with genesis data

kmenon01 committed:
merge conflicts with develop fixed

Ericson2314 committed:
Merge branch 'develop' into create_chain_tx

Ericson2314 committed:
Start defining types for human string names

kmenon01 committed:
parsing most of the transaction, except genesis data, need to work with John to understand error

kmenon01 committed:
minor change

kmenon01 committed:
Finally managed to parse create chain and genesis data is being hashed properly

kmenon01 committed:
merged with develop and ready for PR

kmenon01 committed:
removed .only

patrick-ogrady commented on src/parser.c:
```diff
@@ -485,7 +500,7 @@ enum parse_rv parse_SubnetAuth(struct SubnetAuth_state *const state, parser_meta
       // Type ID
       CALL_SUBPARSER(uint32State, uint32_t);
-      PRINTF("SUBNET AUTH\n");
+      //PRINTF("SUBNET AUTH\n");
```
Should we remove instead of commenting?

patrick-ogrady commented on src/parser.c:
```diff
@@ -520,13 +535,12 @@ enum parse_rv parse_SubnetAuth(struct SubnetAuth_state *const state, parser_meta
         {
           INIT_SUBPARSER(uint32State, uint32_t);
-          RET_IF_PROMPT_FLUSH;
+	  //PRINTF("Reached\n");
```
Remove?

patrick-ogrady commented on src/parser.c:
```diff
@@ -1371,6 +1385,74 @@ enum parse_rv parse_AddSNValidatorTransaction(
+      CALL_SUBPARSER(uint32State, uint32_t);
+      state->state++;
+      state->gen_n = state->uint32State.val;
+      PRINTF("Gen Data Count\n");
```
Debug log?

kmenon01 replied:
Yes this is for the DEBUG log

patrick-ogrady commented on src/parser.c:
```diff
@@ -1371,6 +1385,74 @@ enum parse_rv parse_AddSNValidatorTransaction(
+      INIT_SUBPARSER(uint8State, uint8_t);
+    } fallthrough;
+    case 1: {
+      //
```
Empty comment

patrick-ogrady commented on tests/p-chain-tests.ts:
```diff
@@ -430,16 +431,138 @@ describe('Staking tests', async function () {
-      {header: 'Sign', body: 'Create Subnet'},
-      {header: 'Threshold', body: '1'},
-      {header: 'Address', body: 'local1mg47uqd7stkvqrp57ds7m28txra45u2uzkta8n' },
+      {header: 'Sign', body: 'Create Chain'},
```
This is great

patrick-ogrady commented on src/parser.h:
```diff
@@ -389,7 +444,7 @@ typedef struct {
         output_prompt_t output_prompt;
     } data;
-} prompt_entry_t;
+}prompt_entry_t;
```
Did we mean to remove this space?

kmenon01 committed:
fixed comments

kmenon01 committed:
Merge branch 'develop' into create_chain_tx

patrick-ogrady reviewed (approved):
Nice!

Ericson2314 committed:
Cleanup whitespace

kmenon01 committed:
tests pass with new suite

Ericson2314 committed:
Merge branch 'develop' into create_chain_tx

Ericson2314 committed:
Cleanup whitespace

Ericson2314 committed:
Fix warnings

Ericson2314 committed:
Deduplicate and make easier to read some hashing code

Ericson2314 committed:
Fix up new parsing logic to be more efficient

1. Do not parse "byte by byte" but consume whole chunks as they are
   available. We can drop the 4 loop in this way and just use the
   "async, reentry" patterns without extra mechanics.

2. Do not hash in linear space O(n). Instead hash in constant space
   O(1). Otherwise we will fail when the genesis data is large.

Ericson2314 commented:
@kmenon01 Have a good look at the last commit especially. Without hashing in constant space, we will fail as badly on large genesis data as if were to store and print the entire thing out without hashing it.

Ericson2314 committed:
Fix build error
